lib/evse_security/certificate/x509_hierarchy.cpp

```diff
@@ -121,74 +121,160 @@ std::string X509CertificateHierarchy::to_debug_string() {
-                // We can identify the invalid hash data by the fact that its hash strings are empty
-                CertificateHashData invalid;
-                // Set the hash algorithm, leaving it undefined leads to UB
-                invalid.hash_algorithm = HashAlgorithm::SHA256;
-                ordered.hierarchy.push_back({certif, invalid, certif, {}});
+        // Found a non-permanent orphan, search for a issuer
+        bool found_issuer = false;
+
+        for_each([&](X509Node& top) {
+            if (orphan.certificate.is_child(top.certificate)) {
+                orphan.hash = orphan.certificate.get_certificate_hash_data(top.certificate);
+                orphan.state.is_hash_computed = 1;
+                orphan.state.is_orphan = 0;                   // Not an orphan any more
+                orphan.issuer = X509Wrapper(top.certificate); // Set the new valid issuer
+
+                top.children.push_back(std::move(orphan));
+                found_issuer = true;
             }
-        }
-    });
 
-    // Remove all root certificates from the provided certificate list
-    auto remove_roots = std::remove_if(certificates.begin(), certificates.end(), [&](const X509Wrapper& certif) {
-        return std::find_if(ordered.hierarchy.begin(), ordered.hierarchy.end(), [&](const X509Node& node) {
-                   return (certif == node.certificate);
-               }) != ordered.hierarchy.end();
-    });
+            return (false == found_issuer);
+        });
+
+        if (false == found_issuer) {
+            // Mark as permanent orphan
+            orphan.state.is_orphan = 1; // Permanent orphan
+            orphan.state.is_hash_computed = 0;
+        } else {
+            // Erase from hierarchy list and decrement iterator
+            hierarchy.erase(std::begin(hierarchy) + i);
+            i--;
+        }
+    }
+}
 
-    certificates.erase(remove_roots, certificates.end());
+X509CertificateHierarchy X509CertificateHierarchy::build_hierarchy(std::vector<X509Wrapper>& certificates) {
+    X509CertificateHierarchy ordered;
 
-    // Try build the full hierarchy, we are not assuming any order
-    // This will not get stuck in a loop, because we removed the orphaned certificates already
-    // Note: the logic is fairly simple here, but has worst-case cubic runtime in the number of certs
     while (certificates.size()) {
-        // The current certificate that we're testing for
+        // Remove from back to prevent content moving
         auto current = std::move(certificates.back());
         certificates.pop_back();
 
-        // If we have any roots try and search in the hierarchy the owner
-        if (!ordered.try_add_to_hierarchy(std::move(current))) {
-            // If we couldn't add to the hierarchy move it down in the queue and try again later
-            certificates.insert(certificates.begin(), std::move(current));
-        }
+        ordered.insert(std::move(current));
```

Comment:
Why not move directly into the insert function and pop after that? Then there is no need for the temporary value ;)

Reply:
Just deleted some legacy code, and forgot to properly cleanup the code, no reason except that :)